### PR TITLE
[DebugInfo] Make DIArgList inherit from Metadata and always unique

### PR DESCRIPTION
**Pull request by SLTozer**
This patch changes the `DIArgList` class's inheritance from `MDNode` to `Metadata, ReplaceableMetadataImpl`, and ensures that it is always unique, i.e. a distinct DIArgList should never be produced.

Currently a distinct DIArgList will never be emitted to bitcode or IR, but they could appear in-memory if a DIArgList had one of its arguments RAUWd such that the contents of the DIArgList were identical to an already-existing DIArgList. Adding support for reuniquing DIArgLists in these cases caused some significant performance regressions (~3%); these are resolved by changing DIArgList to no longer inherit from MDNode.

Conceptually, having DIArgList inherit from MDNode does not make a lot of sense - part of the initial reason for the relationship was to make use of MDNode's RAUW support, but in the end the number of exceptions and special cases for DIArgLists makes it probably not worth the shared logic. Outside of the RAUW logic, DIArgLists don't make use of any features of MDNodes, so it makes sense to simply disentangle them.

This should not result in any changes to IR or bitcode; the parsing and printing format for DIArgList is unchanged, and the order in which it appears should also be identical. As a minor note, this patch also fixes a gap in the verifier, where the ValueAsMetadata operands to a DIArgList would not be visited.

### CODE REVIEW

**Comment by dwblaikie** [comment]
> My previous comment still applies: the description should describe the choice of UINT32_MAX and UINT64_MAX, so that a curious reader doesn't have to read several external links to understand the choice.

Is there a way we could avoid hardcoding this, and just do "max of whatever the relocation size is"? (some earlier versions of this patch had switches over relocation types to try to do this - which seemed a bit verbose/brittle/repetitive - but if there was some way to generalize it tidily, I think that'd be ideal)

**Comment by SLTozer** [comment]
@dwblaikie Any chance you meant to comment that on this review? https://github.com/llvm/llvm-project/pull/70701

**Comment by dwblaikie** [comment]
> @dwblaikie Any chance you meant to comment that on this review? #70701

Ah, yep - sorry for the noise. I'll delete my misplaced comment.

**Comment by OCHyams** [comment]
LGTM, sounds like a significant win. I would feel more comfortable if another pair of eyes could look over the changes too though.

**Comment by jmorse** [comment]
Digging into this now, thanks for making this adjustment. Two high level concerns:
 * It looks like some files might end up being different according to the compile time tracker [0], in that one or two of the bitcode files report file-size differences. It's not clear to me whether that's actually a spurious difference though, as the numeric sizes aren't different.
 *  IIRC the reason why https://reviews.llvm.org/D108968 was filed was because of a nondeterminism problem when uniquifying DIArgLists that only appears deep in LTO builds. We should be careful about re-enabling DIArgList uniquification in light of that, I see it's being re-implemented here (will look), is that likely to solve the reported nondeterminism issue?

[0] http://llvm-compile-time-tracker.com/compare.php?from=9e618e5ed4c9f2c24e2c308e7f28e505e788f5fa&to=9552698bada4e3809331d57c0a78627161767ae6&stat=size-file&details=on

File: llvm/lib/Bitcode/Writer/BitcodeWriter.cpp
Comment (jmorse) [inline llvm/lib/Bitcode/Writer/BitcodeWriter.cpp:1982]: For my understanding, what are the implications of not having an abbrev here -- will we end up with a potentially suboptimal encoding or something?

File: llvm/lib/IR/AsmWriter.cpp
Comment (jmorse) [inline llvm/lib/IR/AsmWriter.cpp:1269]: To confirm my understanding: the mechanism of print DIArgLists isn't changing at all, but as they're no longer MDNodes,  they don't get considered for a metadata-node-number slot so don't need to appear here?

File: llvm/lib/IR/Metadata.cpp
Comment (jmorse) [inline llvm/lib/IR/Metadata.cpp:419]: This all sounds good given that there's no performance regression

**Review by jmorse — APPROVED** [review]
Changes LGTM with some understanding-confirmations

**Comment by SLTozer** [comment]
>     * It looks like some files might end up being different according to the compile time tracker [0], in that one or two of the bitcode files report file-size differences. It's not clear to me whether that's actually a spurious difference though, as the numeric sizes aren't different.

I'll try and reproduce this before merging to make sure nothing bad is going on - I have suspicions that this is an innocent case of us being able to merge debug history ranges that previously were pinged as different due to distinct DIArgLists, but there's a (`+0.00%`) size increase for 7zip, so I'll need to figure out whether anything weird is happening there.

>     * IIRC the reason why https://reviews.llvm.org/D108968 was filed was because of a nondeterminism problem when uniquifying DIArgLists that only appears deep in LTO builds. We should be careful about re-enabling DIArgList uniquification in light of that, I see it's being re-implemented here (will look), is that likely to solve the reported nondeterminism issue?

The non-determinism problem solved in that patch was essentially that we _weren't_ properly reuniquing DIArgLists, so we ended up in situations where equal DIArgLists had different pointers. This change should properly reunique DIArgLists, and should actually close a potential case where equal DIArgLists can have different pointers (i.e. if one is made distinct).

File: llvm/lib/Bitcode/Writer/BitcodeWriter.cpp
Comment (SLTozer) [reply llvm/lib/Bitcode/Writer/BitcodeWriter.cpp:1982]: DIArgLists have no abbreviations, so this is a no-op I believe.

File: llvm/lib/IR/AsmWriter.cpp
Comment (SLTozer) [reply llvm/lib/IR/AsmWriter.cpp:1269]: Correct, this function only gets called for MDNodes, so DIArgList doesn't need to care anymore.

**Comment by SLTozer** [comment]
Update, doesn't look like anything serious - I can do some serious digging to verify whether the additional dbg.value produced (which is a rare case) is correct or not, but it is not horribly wrong, and if there is anything incorrect about its value or placement it is almost certainly exposed by this patch, rather than caused by it - does that seem good enough to merge this?

**Comment by jmorse** [comment]
Urgh -- an extra dbg.value from an underlying metadata change shows something is majorly wrong. Although I'd hazard a guess that it isn't this patch, it's something else being exercised. (I'll dig into it).

**Comment by SLTozer** [comment]
> Urgh -- an extra dbg.value from an underlying metadata change shows something is majorly wrong. Although I'd hazard a guess that it isn't this patch, it's something else being exercised. (I'll dig into it).

There is a practical reason to expect a metadata change - prior to this patch, we have the possibility that a DIArgList will be made distinct, and so an equivalent DIArgList that has been properly uniqued will exist at a different address. Since we generally assume DIArgLists are uniqued (as with ValueAsMetadata), all the checks for equality are tests for pointer equality. This change ensures that DIArgLists are actually always uniqued, so value equality will always mean pointer equality, so we may see situations where we correctly identify two locations as being equal when previously we wouldn't. There are probably passes where this results in a debug info change - albeit very rarely, since the distinct case seems to have been quite rare to begin with.

**Comment by jmorse** [comment]
I was a bit skeptical about this ^, after some back and forth with @SLTozer though we reckon it's probably `isIdenticalToWhenDefined` returning something different now when merging blocks in SimplifyCFG, which would result in dbg.values not-being-dropped where they would have been before. With that in mind, LGTM again.

**Comment by SLTozer** [comment]
MemSanitizer buildbot failure, will revert until the problem's fixed.

**Pull request merged**
